tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/jvmmodel/TclJvmModelInferrer.xtend

```diff
@@ -168,48 +193,69 @@ class TclJvmModelInferrer extends AbstractModelInferrer {
+		Iterable<MacroTestStepContext> macroUseStack, Set<EnvParam> envParams) {
 		val mapping = getVariableToValueMapping(step, interaction.template)
-		val values = interaction.defaultMethod.parameters.map [ templateVariable |
+		val stepContents = interaction.defaultMethod.parameters.map [ templateVariable |
 			val stepContent = mapping.get(templateVariable)
-			if (stepContent instanceof StepContentElement) {
-				val element = stepContent.componentElement
-				return element.locator
-			} else if (stepContent instanceof StepContentVariableReference) {
-				return stepContent.dereferenceMacroVariableReference(macroCallStack)
-			} else {
-				return stepContent.value
-			}
+			val stepContentDereferenced = if (stepContent instanceof StepContentVariableReference) {
+					stepContent.dereferenceVariableReference(macroUseStack, envParams)
+				} else {
+					stepContent
+				}
+			return stepContentDereferenced
 		]
 		val typedValues = newArrayList
-		val operationParameters = interaction.defaultMethod.operation.parameters
-		values.forEach [ value, i |
-			val jvmParameter = operationParameters.get(i)
-			if (jvmParameter.parameterType.qualifiedName == String.name) {
-				typedValues += '''"«value»"'''
-			} else {
-				typedValues += value
-			}
+		stepContents.forEach [ stepContent, i |
+			val jvmParameter = interaction.getTypeOfFixtureParameter(i)
+			typedValues += generateCallParameter(stepContent, jvmParameter)
 		]
 		return typedValues.join(', ')
 	}
 
 	/**
-	 * resolve dereferenced variable in macro with call site value (recursively if necessary)
+	 * generate the parameter-code passed to the fixture call depending on the type of the step content
+	 */
+	private def String generateCallParameter(StepContent stepContent, JvmTypeReference expectedType) {
+		switch stepContent {
+			StepContentElement:
+				return '''"«stepContent.componentElement.locator»"'''
+			StepContentVariableReference:
+				if (expectedType.qualifiedName != String.name) {
+					throw new RuntimeException('''Environment variable '«stepContent.variable.name»' (always of type String) is used where type '«expectedType.qualifiedName»' is expected.''')
+				} else {
+					return stepContent.variable.name.envParamToVarName
+				}
+			StepContentValue:
+				if (expectedType.qualifiedName == String.name) {
+					return '''"«stepContent.value»"'''
+				} else {
+					return stepContent.value
+				}
+			default: throw new RuntimeException('''StepContent type ('«stepContent.class.canonicalName»') unknown.''')
+		}
+	}
+
+	/**
+	 * resolve dereferenced variable (in macro) with call site value (recursively if necessary)
 	 */
-	private def String dereferenceMacroVariableReference(StepContentVariableReference referencedVariable,
-		Iterable<MacroTestStepContext> macroCallStack) {
-		val callSiteMacroContext = macroCallStack.head
+	private def StepContent dereferenceVariableReference(StepContentVariableReference referencedVariable,
```

Comment:
It's unclear to me what this method does. Could you add some more comments (JavaDoc and inline) and maybe rename it?